### DISCUSSION TITLE
Clarification on state update and component rerendering

### DISCUSSION THREAD

**Discussion by jalvarezz13** [pmndrs/zustand#2278] [Q&A]
I've implemented a state update in my code, but I'm uncertain whether it will trigger a re-render of components utilizing the 'transactions' value. I've encountered a similar issue while searching for information about this (https://github.com/pmndrs/zustand/issues/244) and suspect the problem might lie in how I'm updating the state.

Here's the code snippet in question:
```ts
addTransaction: async (transaction: ITransaction) => {
  const newTransactions = [...get().transactions, transaction].sort((a, b) => {
    return moment(b.timestamp).diff(moment(a.timestamp))
  })

  set(({ transactions: newTransactions }))
},
```
I'd appreciate any insights or guidance on whether this state update would prompt a re-render of components relying on the 'transactions' data. Thanks in advance for your help!

pd: my store looks like:
```ts
type BlaBlaState = {
  transactions: ITransaction[]
  fetchTransactions: () => void
  addTransaction: (transaction: ITransaction) => void
  updateTransaction: (transaction: ITransaction) => void
}

export const useStore = create(
  persist<BlaBlaState>(
    (set, get) => ({
      transactions: [],
      fetchTransactions: async () => {
        // Some logic to fetch transactions
      },
      addTransaction: async (transaction: ITransaction) => {
        const newTransactions = [...get().transactions, transaction].sort((a, b) => {
          return moment(b.timestamp).diff(moment(a.timestamp))
        })

        set({ transactions: newTransactions })
      },
      updateTransaction: (transaction: ITransaction) => {
        // TODO: Update transaction
      },
    }),
    { name: "bla-bla", storage: createJSONStorage(() => sessionStorage) }
  )
)

```

**Comment by dbritto-dev** [comment]
@jalvarezz13 for async operations you should use previous state directly from `set` function using a callback 

```tsx
addTransaction: async (transaction: ITransaction) => {
  set(state => ({ transactions: state.transactions.concat([transaction]).sort(...) }))
},

```

**Reply by jalvarezz13** [reply]
OK thanks!  But the code you have provided is identical to mine, right?

**Reply by dbritto-dev** [reply]
> OK thanks!  But the code you have provided is identical to mine, right?

@jalvarezz13 yes and no, yes in the sense of we are updating the same field, but no in the sense of how works set state on async operations. You can test it out updating a counter using async operations. 

Ref: https://react.dev/reference/react/useState#updating-state-based-on-the-previous-state
